File: spatial-index/src/main/java/lva/spatialindex/storage/AbstractStorage.java
package lva.spatialindex.storage;

import lombok.SneakyThrows;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;

import static com.google.common.base.Preconditions.checkArgument;

/**
 * @author vlitvinenko
 */
public abstract class AbstractStorage<T> implements Storage<T> {

    protected interface Serializer<T> {
        byte[] serialize(T t);
        T deserialize(byte[] bytes);
    }

    abstract protected static class AbstractSerializer<T> implements Serializer<T> {
        @Override
        @SneakyThrows
        public byte[] serialize(T t) {
            try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
                write(os, t);
                return os.toByteArray();
            }
        }

        @Override
        @SneakyThrows
        public T deserialize(byte[] bytes) {
            try (ByteArrayInputStream is = new ByteArrayInputStream(bytes)) {
                return read(is);
            }
        }

        abstract public void write(OutputStream os, T t);
        abstract public T read(InputStream is);
    }


    private final int recordSize;
    private final StorageSpace storageSpace;

    public AbstractStorage(StorageSpace storageSpace, int recordSize) {
        this.storageSpace = storageSpace;
        this.recordSize = recordSize;
    }

    protected abstract Serializer<T> getSerializer();

    @Override
    public long add(T t) {
        byte[] buff = getSerializer().serialize(t);
        checkArgument(buff.length <= recordSize, "record max size exceeds");

        long offset = storageSpace.allocate(roundToRecordSize(buff.length));
        storageSpace.writeBytes(offset, buff);
        return offset;

    }

    @Override
    public void write(long offset, T t) {
        byte[] buff = getSerializer().serialize(t);
        checkArgument(buff.length <= recordSize, "record max size exceeds");
        checkArgument(offset >= 0, "out of bounds");

        storageSpace.writeBytes(offset, buff);
    }

    @Override
    public T read(long offset) {
        byte[] buff = storageSpace.readBytes(offset, recordSize);
        return getSerializer().deserialize(buff);
    }

    @Override
    public void clear() {
        storageSpace.clear();
    }

    private long roundToRecordSize(long size) {
        return (size + (recordSize - 1)) & -recordSize;
    }
}

File: spatial-index/src/main/kotlin/lva/spatialindex/storage/AbstractStorage.kt
package lva.spatialindex.storage

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.OutputStream

/**
 * @author vlitvinenko
 */
abstract class AbstractStorage<T : Any>(private val storageSpace: StorageSpace, private val recordSize: Int) : Storage<T> {
    interface Serializer<T : Any> {
        fun serialize(t: T): ByteArray
        fun deserialize(bytes: ByteArray): T
    }

    abstract class AbstractSerializer<T : Any> : Serializer<T> {
        override fun serialize(t: T): ByteArray =
            ByteArrayOutputStream().use {
                serializeTo(it, t)
                it.toByteArray()
            }

        override fun deserialize(bytes: ByteArray): T =
            ByteArrayInputStream(bytes).use { deserializeFrom(it) }

        abstract fun serializeTo(outputStream: OutputStream, t: T)
        abstract fun deserializeFrom(inputStream: InputStream): T
    }

    abstract val serializer: Serializer<T>

    override fun add(t: T): Long {
        val bytes = toBytes(t)
        return storageSpace.allocate(roundToRecordSize(bytes.size).toLong())
            .also { storageSpace.writeBytes(it, bytes) }
    }

    override fun write(offset: Long, t: T) {
        check(offset >= 0) { "out of bounds. Offset: $offset" }
        storageSpace.writeBytes(offset, toBytes(t))
    }

    override fun read(offset: Long): T {
        val bytes = storageSpace.readBytes(offset, recordSize)
        return serializer.deserialize(bytes)
    }

    private fun toBytes(t: T): ByteArray = serializer.serialize(t).also {
        check(it.size <= recordSize) { "record max size exceeds. Max size: $recordSize" }
    }

    override fun clear() =
        storageSpace.clear()

    private fun roundToRecordSize(size: Int): Int =
        (size + (recordSize - 1)) and -recordSize
}